plugins/parcelport/verbs/readers_writers_mutex.hpp

```diff
@@ -0,0 +1,264 @@
+namespace lcos {
+namespace local {
+
+    class readers_writer_mutex
+    {
+    private:
+
+        typedef union {
+            uint64_t u;
+            struct {
+                uint32_t wr;        // Writers and readers
+            } i;
+            struct {
+                uint16_t writers;   // Now serving for writers
+                uint16_t readers;   // Now serving for readers
+                uint16_t next;      // Next available ticket number
+                uint16_t notused;   // Padding to 8 bytes
+            } s;
+        } readwrite_ticket;
+
+        readwrite_ticket ticket;
+
+    public:
+        readers_writer_mutex() : ticket{0} {}
+
+        //
+        // acquire lock for a unique writer
+        //
+        void lock()
+        {
+            // memory ordering barrier
+            rwl_barrier_();
+
+            uint16_t val = atomic_xadd(&ticket.s.next, 1);
+            while (val != ticket.s.writers) {
+// std::cout << "rw lock sleep write " << val << " " << ticket.s.next << " \n";
+                hpx::util::detail::yield_k(4, nullptr);
+            }
+
+            // memory ordering rwl_barrier_
+            rwl_barrier_();
+        }
+
+        //
+        // unlock writer
+        //
+        void unlock()
+        {
+            // only one writer can enter unlock at a time, so we do not need atomic ops
+            readwrite_ticket new_ticket = ticket;
+            //
+            ++new_ticket.s.writers;
+            ++new_ticket.s.readers;
+            ticket.i.wr = new_ticket.i.wr;
+        }
+
+        //
+        // try to obtain unique writer lock
+        //
+        bool try_lock()
+        {
+            readwrite_ticket new_ticket, old_ticket;
+            new_ticket = old_ticket = ticket;
+
+            /*
+             * This write lock can only be granted if the lock was last granted to
+             * a writer and there are no readers or writers blocked on the lock,
+             * that is, if this thread's ticket would be the next ticket granted.
+             * Do the cheap test to see if this can possibly succeed (and confirm
+             * the lock is in the correct state to grant this write lock).
+             */
+            if (old_ticket.s.writers != old_ticket.s.next)
+                return false;
+
+            // The replacement lock value is a result of allocating a new ticket.
+            ++new_ticket.s.next;
+
+            return (cmpxchg(&ticket.u, old_ticket.u, new_ticket.u) ? true : false);
+        }
+
+        //
+        // obtain a reader lock, many readers may have the lock simultaneously
+        //
+        void lock_shared()
+        {
+            // memory ordering rwl_barrier_
+            rwl_barrier_();
+
+            uint16_t val = atomic_xadd(&ticket.s.next, 1);
+            while (val != ticket.s.readers) {
+// std::cout << "rw lock sleep read " << val << " " << ticket.s.readers << "\n";
+                hpx::util::detail::yield_k(0, nullptr);
```

Comment:
You should use `yield_k as:

    for (std::size_t k = 0; val != ticket.s.readers; ++k)
    {
        hpx::detail::yield_k(k);
    }